### PR TITLE
Improve `dolfinx.common` documentation

### PR DESCRIPTION
**Pull request by jorgensd**
Expose various functions through `__all__`.

Update docs for timings, and add more documentation.

Add documentation for user-facing IndexMap.

### CODE REVIEW

File: python/dolfinx/common.py [hunk 1]
```diff
@@ -5,35 +5,47 @@
...
+    """Print out a summary of all Timer measurements.
+
+    One can specify the timing types as a subset of wall time, system time or user time.
+
+    Note:
+        When used in parallel, a reduction is applied across all processes. By default, the maximum time is shown.
+
+    Args:
+        comm: Communicator to reduce timings over
+        timing_types: List of timings to return
```
Comment (garth-wells) [inline python/dolfinx/common.py:39]: What's the type (e.g., list of enums?) and what are the possible values?

File: python/dolfinx/common.py [hunk 2]
```diff
@@ -5,35 +5,47 @@
...
 
-def list_timings(comm, timing_types: list, reduction=Reduction.max):
-    """Print out a summary of all Timer measurements, with a choice of
-    wall time, system time or user time. When used in parallel, a
-    reduction is applied across all processes. By default, the maximum
-    time is shown."""
+def list_timings(comm: _MPI.Comm, timing_types, reduction=Reduction.max):
+    """Print out a summary of all Timer measurements.
+
+    One can specify the timing types as a subset of wall time, system time or user time.
```
Comment (garth-wells) [inline python/dolfinx/common.py:32]: I think this is redundant if `timing_types` is expanded on.

File: python/dolfinx/common.py [hunk 3]
```diff
@@ -91,7 +102,17 @@ def elapsed(self):
 
 
 def timed(task: str):
-    """Decorator for timing functions."""
+    """Decorator for timing functions. The basic usage is::
+
+        @timed('Description of function')
```
Comment (garth-wells) [inline python/dolfinx/common.py:107]: Google doctring style uses `Examples:` section for code example usage, see https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html.

File: python/dolfinx/wrappers/common.cpp [hunk 4]
```diff
@@ -145,16 +161,32 @@ void common(nb::module_& m)
             return std::pair(std::move(map),
                              dolfinx_wrappers::as_nbarray(std::move(ghosts)));
           },
-          nb::arg("entities"));
+          nb::arg("entities"),
+          nb::raw_doc("Create new index map from a subset of indices in this "
```
Comment (garth-wells) [inline python/dolfinx/wrappers/common.cpp:165]: This is too fragile. The approach we've adopted (which is not fully complete) is to document from Python.

File: python/dolfinx/common.py (hunk 1 above)
Comment (jorgensd) [reply python/dolfinx/common.py:39]: Could not use type hints due to nanobind missing stubs (so mypy failed).
Can add a reference to the cpp documentation

File: python/dolfinx/wrappers/common.cpp (hunk 4 above)
Comment (jorgensd) [reply python/dolfinx/wrappers/common.cpp:165]: So then we need to wrap every function we expose to the Python interface?

**Pull request closed without merging**